File: J9Ch/src/S_5_Dynamic_Problem_I/Required_10/_398_Longest_Increasing_Continuous_subsequence_II_Hard.java
package S_5_Dynamic_Problem_I.Required_10;
import java.util.*;
/** 398 Longest Increasing Continuous subsequence II


 * Created by tianhuizhu on 6/28/17.
 */
public class _398_Longest_Increasing_Continuous_subsequence_II_Hard {

    public class Solution {
        /**
         * @param A an integer matrix
         * @return  an integer
         */
        int [][]dp;
        int [][]flag ;
        int n ,m;
        public int longestIncreasingContinuousSubsequenceII(int[][] A) {
            // Write your code here
            if(A.length == 0)
                return 0;
            n = A.length;
            m  = A[0].length;
            int ans= 0;
            dp = new int[n][m];
            flag = new int[n][m];

            for(int i = 0; i < n; i++) {
                for(int j = 0; j < m; j++) {
                    dp[i][j] = search(i, j, A);
                    ans = Math.max(ans, dp[i][j]);
                }
            }
            return ans;
        }
        int []dx = {1,-1,0,0};
        int []dy = {0,0,1,-1};

        int search(int x, int y, int[][] A)   {
            if(flag[x][y] != 0)
                return dp[x][y];

            int ans = 1;
            int nx , ny;
            for(int i = 0; i < 4; i++) {
                nx = x + dx[i];
                ny = y + dy[i];
                if(0<= nx && nx < n && 0<= ny && ny < m ) {
                    if( A[x][y] > A[nx][ny]) {
                        ans = Math.max(ans,  search( nx, ny, A) + 1);
                    }
                }
            }
            flag[x][y] = 1;
            dp[x][y] = ans;
            return ans;
        }
    }
}

File: J9Ch/src/S_5_Dynamic_Problem_I/Required_10/python_cpp/_398_Longest_Increasing_Continuous_subsequence_II_Hard.cpp

const int dirx[4] = {0, 0, 1, -1};
const int diry[4] = {1, -1, 0, 0};
class Solution {
public:
    /**
     * @param A an integer matrix
     * @return  an integer
     */
    int n, m;
    int dis[200][200];
    int maxx;
    void dfs(int x,int y,int depth, vector<vector<int>>& a) {
        int nowdepth = depth;
            if ((++nowdepth) > dis[x][y]) {
                dis[x][y] = nowdepth;
                if (dis[x][y] > maxx) maxx = dis[x][y];
        }
        else
            return;
        int nowx,nowy;
        for (int i=0;i < 4; ++i) {
            nowx = x + dirx[i];
            nowy = y + diry[i];
            if (nowx < 0 || nowx >= n || nowy < 0 || nowy >= m) continue;
            if (a[nowx][nowy] <= a[x][y]) continue;
            dfs(nowx, nowy, nowdepth, a);
        }
    }
    int longestIncreasingContinuousSubsequenceII(vector<vector<int>>& A) {
        // Write your code here
        maxx = 0;
        n = A.size();
        if (n == 0)
            return 0;
        m = A[0].size();
        if (m == 0)
            return 0;
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < m; ++j)
                dis[i][j] =0;

        for (int i = 0; i < n; ++i)
            for (int j = 0; j < m; ++j)
                if (dis[i][j] == 0)
                    dfs(i, j , 0, A);
        return maxx;
    }
};